controllers/machinedeletionremediation_controller.go

```diff
@@ -76,73 +99,125 @@ type MachineDeletionRemediationReconciler struct {
+		}
 
-	if !hasControllerOwner(machine) {
-		log.Info("ignoring remediation of node-associated machine: the machine has no controller owner", "machine", machine.GetName(), "node name", remediation.Name)
-		return ctrl.Result{}, nil
-	}
+		err = r.updateConditions(remediationStarted, remediation)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+	case mdrStateStarted:
+		// request deletion
+		// set machine deletion annotation
+		node, machine, err := r.getNodeAndMachineFromCR(remediation)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
 
-	log.Info("node-associated machine found", "node", node.Name, "machine", machine.GetName())
+		log.Info("requesting node-associated machine deletion", "node", node.GetName(), "machine", machine.GetName())
+		err = r.deleteMachineOfNode(ctx, machine, remediation.Name)
+		if err != nil {
+			updateErr := r.updateConditions(remediationFailed, remediation)
+			if updateErr != nil {
+				err = errors.Wrapf(err, "%v", updateErr)
+			}
+			return ctrl.Result{}, err
+		}
 
-	err = r.deleteMachineOfNode(ctx, machine, remediation.Name)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
+		err = r.setDeletedMachineAnnotation(ctx, remediation, machine)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrapf(err, "failed to update remediation CR with timeout annotation")
+		}
 
-	err = r.setDeletedMachineAnnotation(ctx, remediation, machine)
-	if err != nil {
-		return ctrl.Result{}, errors.Wrapf(err, "failed to update remediation CR with timeout annotation")
+		// requeue immediately to check machine deletion progress
+		finalResult = ctrl.Result{Requeue: true}
+
+	case mdrStateDeletionRequested:
+		// follow machine deletion progress
+		// set status.condition RemediationFinished/RemediationFailed
+		annotations := remediation.GetAnnotations()
+		if annotations != nil {
+			machineNameNamespace, exists := annotations[MachineNameNamespaceAnnotation]
+			if exists {
+				machineName, machineNamespace, err := extractNameAndNamespace(machineNameNamespace, remediation.GetName())
+				if err != nil {
+					log.Error(err, "could not get Machine data from remediation", "remediation", remediation.GetName(), "annotation", machineNameNamespace)
+					return ctrl.Result{}, r.updateConditions(remediationFailed, remediation)
+				}
+
+				ok, err := r.verifyMachineIsDeleted(ctx, machineName, machineNamespace, remediation.Name)
+				if err != nil {
+					return ctrl.Result{}, r.updateConditions(remediationFailed, remediation)
+				}
+				if !ok {
+					return ctrl.Result{RequeueAfter: 30 * time.Second}, nil
+				}
+
+				return ctrl.Result{}, r.updateConditions(remediationFinished, remediation)
+			}
+		}
+	// For completeness, we handle failed and completed switch-case here.
+	// Any errors was already propagated, so we can handle them the same way,
+	// returning any possible error generated above, even though err == nil is
+	// expected at this point.
+	case mdrStateFailed:
+		fallthrough
+	case mdrStateCompleted:
```

Comment:
I believe you can simply add comma between them
```suggestion
	case mdrStateFailed, mdrStateCompleted:
```

Reply:
And nothing is done for both of these cases? No need to update status or something else?

Reply:
You're right about the `case` switch, I'll update it.

About the two cases, also correct there is nothing to do, since these states come from a "last action" for the operator and the status was already updated in the deferred function.